File: LeetCode/cpp/SplitArrayLargestSum.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <algorithm> 

using namespace std;

class Solution {
public:
    int splitArray(vector<int>& nums, int m) {
        int left = nums[0];
        long long int right = 0;
        int n = nums.size();
        for(int i=0; i<=n; i++)
        {
            right += nums[i];
            left = max(left, nums[i]);
        }
        
        while(left <= right)
        {
            long long int mid = left + (right - left) / 2;
            int count = 0;
            int s = 0;
            for(int i=0; i<n; i++)
            {
                if(s + nums[i] > mid)
                {
                    count++;
                    s = nums[i];
                }
                else
                    s += nums[i];
            }
            if(s > 0)
                count++;
            
            if(count <= m)
                right = mid - 1;
            else
                left = mid + 1;
        }
        return left;
    }
};

int main()
{
    vector<int> nums = {7,2,5,10,8};
    int m = 2;
    Solution s;
    cout << s.splitArray(nums, m) << endl;
    return 0;
}

File: LeetCode/python/SplitArrayLargestSum.py
"""
给定一个非负整数数组和一个整数 m，你需要将这个数组分成 m 个非空的连续子数组。设计一个算法使得这 m 个子数组各自和的最大值最小。

给定nums = [7,2,5,10,8], m = 2, 有4种方法将nums分为2组, 而和最小的情况是[7,2,5] 和 [10,8]

思路: 和最小的值肯定属于[max(nums), sum(nums)], 所以使用二分查找
"""


def cal_split_times(nums, mid):
    res = 0
    s = 0
    i = 0
    while i < len(nums):
        if s + nums[i] <= mid:
            s += nums[i]
            i += 1
        else:
            s = 0
            res += 1
    if s > 0:
        res += 1
    return res


def split_array(nums, m):
    left = max(nums)
    right = sum(nums)
    while left <= right:
        mid = (left + right) // 2
        splits = cal_split_times(nums, mid)
        if splits <= m:
            right = mid - 1
        else:
            left = mid + 1
    return left


if __name__ == "__main__":
    nums = [7, 2, 5, 10, 8]
    m = 2
    print(split_array(nums, m))